docs/architecture/adr-021-restricted-block-size.md

```diff
@@ -0,0 +1,156 @@
+# ADR 021: Restricted and Configurable Block Size
+
+## Changelog
+
+- 2023-05-12: Initial
+
+## Status
+
+Accepted
+
+## Context
+
+Currently, block sizes are controlled by three values, the smallest of which
+will determine what the protocol considers to be a valid block size.
+
+| Parameter        | Size   | Unit   | Description  | Value Control |
+|------------------|--------|--------|--------------|---------------|
+| `MaxBlockSizeBytes` | 100 | MiB | Maximum total size of the protobuf encoded block, a hard coded constant acting as a cap for `MaxBytes`. | Hard coded |
+| `MaxBytes`     |  21 | MiB | Determines the valid size of the entire protobuf encoded block. Is a governance modifiable parameter and is capped by `MaxBytes`. Used to regulate the amount of data gossiped in the consensus portion of the network (using the current block gossiping mechanism). | Modifiable |
+| `MaxSquareSize` | ~7.5 | MiB | Determines the maximum size of the original data square. Used to regulate storage requirements for celestia-node, and the size of the data availability header. Default set to 128. | Hard coded |
+
+Using the currently set/default values, `MaxSquareSize` is limiting the block
+size because it is the smallest value in terms of bytes. However if the
+community changed `MaxBytes` to something below 7.5MiB, then that would become
+the limit on blocksize.
+
+The system is missing the ability for governance to limit the square size, which
+is something that we want in order to reduce the amount of data stored by the
+data availability portion of the network. This could be done now by starting new
+networks with `MaxSquareSize` set to 64. However, this would mean that even
+after a solution is implemented for pruning, a hardfork would be required to
+increase the value back to the current `MaxSquareSize`.
+
+### Reduce the `MaxBytes` Governance Parameter
+
+One solution is to reduce the `MaxBytes` governance parameter to roughly the
+size of a 64 x 64 square (~1.8 MB). While tendermint will reject blocks that
+exceed `MaxBytes`, due to encoding overhead, it's possible for a significantly
+smaller block to require a square size larger than 64. This results in only ever
+confidently achieving a soft block of squares over size 64. Implementation and
+fuzzing test for this option can be found at
+[#1743](https://github.com/celestiaorg/celestia-app/pull/1743).
+
+### Introduce a new Governance Parameter
+
+The second suggestion is to create a new governance parameter,
+`GovMaxSquareSize`, which governance could set. Now the protocol would use the
+lowest of 4 values to determine the size of a block at a given height. This has
```

Comment:
> Now the protocol would use the
lowest of 4 values to determine the size of a block at a given height

[no change needed] I interpret `MaxBlockSizeBytes` as a maximum value for `MaxBytes`. Similarly, I interpret `MaxSquareSize` as a maximum value for `GovMaxSquareSize`.

When evaluating if a square is valid, I expect that the proposed block only needs to be compared with `MaxSquareSize` and `GovMaxSquareSize` and doesn't need to be compared with the caps for those values because the param proposal that attempts to change `MaxBytes` or `GovMaxSquareSize` above their respective caps should fail.

Reply:
tendermint will check first using both the bytes params no matter what, and then we check using the gov param in prepare/process proposal. So we actually end up using the lowest value of all 4 each height